src/lib/storage/chunk.cpp

```diff
@@ -1,31 +1,67 @@
-#include "chunk.hpp"
+#include <boost/preprocessor/seq/for_each.hpp>
 
 #include "abstract_segment.hpp"
+#include "all_type_variant.hpp"
+#include "chunk.hpp"
 #include "utils/assert.hpp"
+#include "value_segment.hpp"
 
 namespace opossum {
 
 void Chunk::add_segment(const std::shared_ptr<AbstractSegment> segment) {
-  // Implementation goes here
+  // FIXME: shared_ptr is copied each time, we could remove the `const` and would technically not change the
+  //        interface, but not sure if upstream is okay with that
+  _segments.emplace_back(segment);
+}
+
+template <typename T>
+static bool try_to_append_to_concrete_value_segment(AbstractSegment* segment, const AllTypeVariant& value) {
+  const auto concrete_value_segment = dynamic_cast<ValueSegment<T>*>(segment);
+  if (concrete_value_segment == nullptr) {
+    return false;
+  }
+  concrete_value_segment->append(value);
+  return true;
 }
 
 void Chunk::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
+  DebugAssert(values.size() == _segments.size(), "Tried to append row with unfitting number of columns.");
+
+  for (size_t i = 0, size = values.size(); i < size; ++i) {
+    // Try all possible instantiations of `ValueSegment`, because we cannot know which subclass is inside the
+    // `AbstractSegment` and still want to use the casting functionality of `ValueSegment`, that is, we cannot assume
+    // that the type of `value` matches this segment.
+
+    const auto& segment = _segments[i];
+    const auto& value = values[i];
+
+#define TRY_WITH_TYPE(_, __, type)                                                                 \
+  {                                                                                                \
+    const bool type_matched = try_to_append_to_concrete_value_segment<type>(segment.get(), value); \
+    if (type_matched) {                                                                            \
+      continue;                                                                                    \
```

Comment:
The `continue` applies to the outer loop, so we continue with the next column instead of trying the other types - and we also avoid reaching the `Fail(...)`. Note that the `BOOST_PP_SEQ_FOR_EACH` does not emit a loop but only expands the given macro for each of the types, so the expanded code would look something like this:
```cpp
for (size_t i = 0, size = values.size(); i < size; ++i) {
  const auto& segment = _segments[i];
  const auto& value = values[i];
  {
    const bool type_matched = try_to_append_to_concrete_value_segment<std::string>(segment.get(), value);
    if (type_matched) {
      continue;
    }
  }
  {
    const bool type_matched = try_to_append_to_concrete_value_segment<float>(segment.get(), value);
    if (type_matched) {
      continue;
    }
  }
  // ...

  Fail("Could not append to chunk, because no concrete segment type was found to append to.");
}
```

Regarding multiple casts succeeding: _If_ we didn't have the continue, this _could_ be a problem, but since we are only checking for `ValueSegment<T>`, it is not possible (and I think it is safe to assume that for different types `T` and `U`, the types `ValueSegment<T>` and `ValueSegment<U>` do not overlap)

Reply:
Oh, right... I haven't really gotten used to thinking in this code expansion semantic of macros, it looks like. 

With the `continue` referring to the outer loop, this actually makes perfect sense here.